tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -367,20 +481,76 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
       }
     }
 
-    std::vector<HloInstruction *> operands(matmul->operands().begin(),
-                                           matmul->operands().end());
-    operands.push_back(bias);
-
-    std::unique_ptr<HloInstruction> fused_op =
-        matmul->CloneWithNewOperands(instr->shape(), operands);
+    // When slicing is applied to the GEMM, replace
+    // add(slice(gemm), broadcast_bias) with
+    // slice(add(gemm, broadcast_bias)) to enable fusing.
+    if (slice) {
+      TF_RETURN_IF_ERROR(
+          SinkSliceOrBitcastBelowBinaryOp(slice, &add, &broadcast_bias));
+      bias = broadcast_bias->mutable_operand(0);
+    }
 
+    // Replace add(gemm, broadcast_bias) with fused new_gemm.
     config.set_epilogue(GemmBackendConfig::BIAS);
-    TF_RETURN_IF_ERROR(fused_op->set_backend_config(config));
-    TF_RETURN_IF_ERROR(SetName(instr->GetModule(), fused_op.get()));
-    TF_RETURN_IF_ERROR(ReplaceWithNewInstruction(instr, std::move(fused_op)));
+    gemm->AppendOperand(bias);
+
+    std::unique_ptr<HloInstruction> new_gemm = HloInstruction::CreateCustomCall(
+        add->shape(), gemm->operands(), kCublasLtMatmulCallTarget);
+    TF_RETURN_IF_ERROR(new_gemm->set_backend_config(config));
+    TF_RETURN_IF_ERROR(SetName(add->GetModule(), new_gemm.get()));
+    TF_RETURN_IF_ERROR(ReplaceWithNewInstruction(add, std::move(new_gemm)));
+
     return true;
   }
 
+  Status FuseReluActivation(HloInstruction *maximum,
+                            HloInstruction *broadcast_zeros,
+                            HloInstruction *gemm,
+                            HloInstruction *slice_or_bitcast = nullptr) {
+    TF_RET_CHECK(ShapeUtil::Compatible(
+        broadcast_zeros->shape(),
+        (slice_or_bitcast ? slice_or_bitcast->shape() : gemm->shape())));
+    auto out_type = gemm->shape().element_type();
+    // Verify that the data type is supported by Epilogue Fusion.
+    if (!SupportsEpilogueFusion(out_type)) {
+      return OkStatus();
+    }
+    bool valid_fusion_pattern =
+        (gemm->operand_count() == 3)
+            ? gemm->operand(0)->shape() != gemm->operand(2)->shape()
```

Comment:
When fusing relu, why must operand 0 and 2 be different shapes?

Reply:
I think it should be fine if a matrix bias is being fused, which for square operands will have the same size as the operands.

Reply:
I don't understand what you're trying to say. You're saying this check can be removed?

Reply:
Yes, I think it can be removed. We were initially planning to only fuse the addition of a vector bias with ReLU, and this seems to be a remnant from that stage.